src/vanilla.ts

```diff
@@ -49,167 +48,195 @@ const snapshotCache = new WeakMap<
+        if (hasPrevValue && customObjectIs(prevValue, value)) {
+          return true
+        }
+        const childListeners = prevValue?.[LISTENERS]
+        if (childListeners) {
+          childListeners.delete(popPropListener(prop))
+        }
+        if (isObject(value)) {
+          value = getUntracked(value) || value
+        }
+        let nextValue: any
+        if (Object.getOwnPropertyDescriptor(target, prop)?.set) {
+          nextValue = value
+        } else if (value instanceof Promise) {
+          nextValue = value
+            .then((v) => {
+              nextValue[PROMISE_RESULT] = v
+              notifyUpdate(['resolve', [prop], v])
+              return v
+            })
+            .catch((e) => {
+              nextValue[PROMISE_ERROR] = e
+              notifyUpdate(['reject', [prop], e])
+            })
+        } else if (value?.[LISTENERS]) {
+          nextValue = value
+          nextValue[LISTENERS].add(getPropListener(prop))
+        } else if (customCanProxy(value)) {
+          nextValue = proxyFunction(value)
+          nextValue[LISTENERS].add(getPropListener(prop))
+        } else {
+          nextValue = value
+        }
+        Reflect.set(target, prop, nextValue, receiver)
+        notifyUpdate(['set', [prop], value, prevValue])
         return true
-      }
-      const childListeners = prevValue?.[LISTENERS]
-      if (childListeners) {
-        childListeners.delete(popPropListener(prop))
-      }
-      if (isObject(value)) {
-        value = getUntracked(value) || value
-      }
-      let nextValue: any
-      if (Object.getOwnPropertyDescriptor(target, prop)?.set) {
-        nextValue = value
-      } else if (value instanceof Promise) {
-        nextValue = value
-          .then((v) => {
-            nextValue[PROMISE_RESULT] = v
-            notifyUpdate(['resolve', [prop], v])
-            return v
-          })
-          .catch((e) => {
-            nextValue[PROMISE_ERROR] = e
-            notifyUpdate(['reject', [prop], e])
-          })
-      } else if (value?.[LISTENERS]) {
-        nextValue = value
-        nextValue[LISTENERS].add(getPropListener(prop))
-      } else if (this.canProxy(value)) {
-        nextValue = proxy(value)
-        nextValue[LISTENERS].add(getPropListener(prop))
+      },
+    }
+    const [baseObject, proxyObject] = customCreateProxy(initialObject, handler)
+    proxyCache.set(initialObject, proxyObject)
+    Reflect.ownKeys(initialObject).forEach((key) => {
+      const desc = Object.getOwnPropertyDescriptor(
+        initialObject,
+        key
+      ) as PropertyDescriptor
+      if (desc.get || desc.set) {
+        Object.defineProperty(baseObject, key, desc)
       } else {
-        nextValue = value
+        proxyObject[key as keyof T] = initialObject[key as keyof T]
       }
-      Reflect.set(target, prop, nextValue, receiver)
-      notifyUpdate(['set', [prop], value, prevValue])
-      return true
-    },
+    })
```

Comment:
Could be simplified by copying of keys in `Reflect.ownKeys(initialObject)` into the default `createProxy` method, which also allows developer to decide which keys from initialObject to keep.

```ts
const proxyObject = customCreateProxy(initialObject, handler);
proxyCache.set(initialObject, proxyObject);
return proxyObject;


const createProxy = <T extends object>(
  initialObject: T,
  handler: ProxyHandler<T>
): readonly [T, T] => {
  const baseObject = Array.isArray(initialObject)
    ? []
    : Object.create(Object.getPrototypeOf(initialObject))
  const proxyObject = new Proxy(baseObject, handler)

  Reflect.ownKeys(initialObject).forEach((key) => {
    const desc = Object.getOwnPropertyDescriptor(
      initialObject,
      key
    ) as PropertyDescriptor
    if (desc.get || desc.set) {
      Object.defineProperty(baseObject, key, desc)
    } else {
      proxyObject[key as keyof T] = initialObject[key as keyof T]
    }
  })
  return proxyObject
}
```

Reply:
I'm not sure if I understand your comment.

Reply:
I’m suggesting to move the for loop which copies keys from the initial object into the proxy object into the default createProxy function. 

So developers can also fully customize which keys to copy if they decide not to use initialObject.

Reply:
Ah, I probably see your point.
I have considered that, and for this time, I decided not to add it after the consideration.
If we got use cases require such capability, we consider it in a follow up PR.

(after all, it's the balance between customization and forking.)

Reply:
Hm, maybe I can do that in this PR. Let me try.